### PR TITLE
makeSetupHook: support depsTargetTargetPropagated (fix using hooks when strictDeps)

### PR DESCRIPTION
**Pull request by Artturin**
Closes https://github.com/NixOS/nixpkgs/issues/49132

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix [hunk 1]
```diff
@@ -65,6 +72,15 @@ makeSetupHook {
         ''
       );
 
+      basic-contains-gdk-pixbuf = let
+        tested = basic;
+      in testLib.runTest "basic-contains-gdk-pixbuf" (
+        testLib.skip stdenv.isDarwin ''
+          ${expectSomeLineContainingYInFileXToMentionZ "${tested}/bin/foo" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
+          ${expectSomeLineContainingYInFileXToMentionZ "${tested}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
```
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:80]:
```suggestion
      basic-contains-gdk-pixbuf = testLib.runTest "basic-contains-gdk-pixbuf" (
        testLib.skip stdenv.isDarwin ''
          ${expectSomeLineContainingYInFileXToMentionZ "${basic}/bin/foo" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
          ${expectSomeLineContainingYInFileXToMentionZ "${basic}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
```

File: pkgs/stdenv/generic/setup.sh [hunk 2]
```diff
@@ -855,6 +855,39 @@ stripHash() {
 }
 
 
+propagatePropagatedDependencies() {
```
Comment (jtojnar) [inline pkgs/stdenv/generic/setup.sh:858]: Maybe something like `{persist,record,output}PropagatedDependencies` so that we are not propagating what is already propagated?

File: pkgs/stdenv/generic/setup.sh [hunk 3]
```diff
@@ -1200,35 +1233,7 @@ fixupPhase() {
 
 
     # Propagate dependencies & setup hook into the development output.
-    declare -ra flatVars=(
```
Comment (jtojnar) [inline pkgs/stdenv/generic/setup.sh:1203]: Rest of the `fixupPhase` function does not seem to use any of these variables (`substituteAll` only captures environment variables) so this looks good to me.

File: pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix [hunk 4]
```diff
@@ -27,16 +28,22 @@ makeSetupHook {
...
-    # graphics in GTK (e.g. cross for closing window in window title bar)
-    # so it is pretty much required for applications using GTK.
     librsvg
   ] ++ [
 
     # We use the wrapProgram function.
     makeWrapper
   ];
+
+  # depsTargetTargetPropagated will be buildInputs
```
Comment (jtojnar) [inline pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:38]: Was this meant to be something like

```suggestion
  # depsTargetTargetPropagated will essentially be buildInputs when wrapGAppsHook is placed into nativeBuildInputs
```

File: pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix [hunk 5]
```diff
@@ -27,16 +28,22 @@ makeSetupHook {
...
+  # the librsvg above should be removed but kept to not break anything that implicitly depended on its binaries
```
Comment (jtojnar) [inline pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:39]: Maybe move the comment above then. I was quite confused until I noticed it.

Personally, I would just rip the band-aid off. I would not expect this to be very prevalent.

File: pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix [hunk 6]
```diff
@@ -27,16 +28,22 @@ makeSetupHook {
...
+  depsTargetTargetPropagated = assert (lib.assertMsg (!targetPackages ? raw) "wrapGAppsHook must be in nativeBuildInputs"); lib.optionals isGraphical [
+    # librsvg provides a module for gdk-pixbuf to allow rendering
+    # SVG icons. Most icon themes are SVG-based and so are some
+    # graphics in GTK (e.g. cross for closing window in window title bar)
+    # so it is pretty much required for applications using GTK.
+    librsvg
```
Comment (jtojnar) [inline pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:45]: This should probably go to a separate commit from “makeSetupHook: support depsTargetTargetPropagated”

File: pkgs/build-support/trivial-builders.nix [hunk 7]
```diff
@@ -559,10 +559,22 @@ rec {
    *                 passthru.tests.greeting = callPackage ./test { };
    *               } ./myscript.sh;
    */
-  makeSetupHook = { name ? "hook", deps ? [], substitutions ? {}, meta ? {}, passthru ? {} }: script:
+  makeSetupHook =
+    { name ? "hook"
+    , substitutions ? { }
+    , meta ? { }
+    , passthru ? { }
+    , deps ? [ ]
```
Comment (jtojnar) [inline pkgs/build-support/trivial-builders.nix:567]: Why change the ordering? It does not appear to be alphabetical nor does it follow the [common order for `mkDerivation`](https://github.com/jtojnar/nixpkgs-hammering/blob/1b038ef38fececb39b65a4cdfa7273ed9d9359b4/explanations/attribute-ordering.md).

File: pkgs/build-support/trivial-builders.nix [hunk 8]
```diff
@@ -559,10 +559,22 @@ rec {
...
+    # hooks go in nativeBuildInput so these will be nativeBuildInput
+    , propagatedBuildInputs ? [ ]
+    # these will be buildInputs
+    , depsTargetTargetPropagated ? [ ]
+    }: script:
```
Comment (jtojnar) [inline pkgs/build-support/trivial-builders.nix:572]: Improve legibility:

```suggestion
    }:
    script:
```

File: pkgs/build-support/trivial-builders.nix [hunk 9]
```diff
@@ -559,10 +559,22 @@ rec {
...
     runCommand name
       (substitutions // {
         inherit meta;
+        inherit depsTargetTargetPropagated;
+        propagatedBuildInputs = assert (lib.assertMsg (lib.isList deps) "deps argument to makeSetupHook must be a list. content: ${toString deps}"); deps ++ propagatedBuildInputs;
```
Comment (jtojnar) [inline pkgs/build-support/trivial-builders.nix:577]: I would go for a soft deprecation (using a `trace`) first so that external consumers can adapt.

And while at it we might as well deprecate `deps` in favour of `propagatedBuildInputs`.

**Review by jtojnar — COMMENTED** [review]
Nice work.

File: pkgs/stdenv/generic/setup.sh (hunk 2 above)
Comment (Artturin) [reply pkgs/stdenv/generic/setup.sh:858]: changed to `recordPropagatedDependencies`

File: pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:79]: @jtojnar what do you think? since you added these tests

File: pkgs/build-support/trivial-builders.nix (hunk 9 above)
Comment (Artturin) [reply pkgs/build-support/trivial-builders.nix]: > I would go for a soft deprecation (using a trace) first so that external consumers can adapt.

done

Comment (Artturin) [reply pkgs/build-support/trivial-builders.nix]: and deps deprecated

File: pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix (hunk 5 above)
Comment (Artturin) [reply pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:38]: > Maybe move the comment above then. I was quite confused until I noticed it.

this comment and the librsvg above it are closer now after rearranging the list

> Personally, I would just rip the band-aid off. I would not expect this to be very prevalent.

Should be a separate PR

**Comment by winterqt** [comment]
Should every instance of e.g. Cargo and pip be from buildPackages, as makeWrapper is?

**Comment by Artturin** [comment]
> Should every instance of e.g. Cargo and pip be from buildPackages, as makeWrapper is?

No, thats just a workaround for hooks in buildInputs / hook in nativeBuildInputs but splicing not working
